### ISSUE TITLE
Feature request: Ignore nanoseconds in datetime strings

### ISSUE THREAD

**Issue by juergenhoetzel** [jcrist/msgspec#200]
First of all: Thanks for this awesome project.

I'm working on a `pyhon` implementation of the [restic](https://restic.net/) backup solution: [pyrrhic](https://github.com/juergenhoetzel/pyrrhic) 

After switching from a naive `pydantic` + stdlib `json` implementation to `msgspec` the performance for loading the backup index improved by a factor of 10x :rocket:  and is now even faster than the original `golang` implementation:  

 The only  downside was I had to add an additional Class and encoder/decoder for ` restic` timestamps which contain nanoseconds like `"2015-01-02T18:10:50.895208559+01:00"`:

```python
DATETIME_TRAILING_NS_RE = re.compile(r"(\.[0-9]{6})[0-9]*\+")


def datetime_from_restic(s: str) -> datetime:
    return datetime.fromisoformat(DATETIME_TRAILING_NS_RE.sub(r"\1+", s))


class resticdatetime(datetime):
    @classmethod
    def fromresticformat(cls, date_string):
        """Construct a datetime from the restic datetime str (strip nanoseconds)."""
        date_string = DATETIME_TRAILING_NS_RE.sub(r"\1+", date_string)
        return cls.fromisoformat(date_string)


def resticdatetime_enc_hook(obj: Any) -> Any:
    if isinstance(obj, resticdatetime):
        return obj.isoformat()
    else:
        raise TypeError(f"Objects of type {type(obj)} are not supported")


def resticdatetime_dec_hook(type: Type, obj: Any) -> Any:
    if type is resticdatetime:
        return resticdatetime.fromresticformat(obj)
    else:
        raise TypeError(f"Objects of type {type} are not supported")

```
to prevent
```
ValidationError: Invalid RFC3339 encoded datetime - at `$.time`
```

Because timestamps in `JSON` are not standardized discarding (optionally?)  nanoseconds would be a nice feature IMHO.

**Comment by jcrist** [comment]
Merged! I'm hoping to hold off releasing until CPython 3.11 is released (2022-10-24), but for now you can install from github to try things out.

```
pip install git+https://github.com/jcrist/msgspec.git
```

Please let me know if/how the changes work out for you, or if you run into any other issues.

**Comment by juergenhoetzel** [comment]
> ```
> 
> ```
> 
> 
>     
>   
> 
> Please let me know if/how the changes work out for you, or if you run into any other issues.

works as expected and makes my implementation much simpler! Again: thanks a lot!
Parsing 500MB of AES-Encrypted index data here:

```
juergen@shaun:~/ghq/github.com/juergenhoetzel/pyrrhic → time pyrrhic ls latest
/var
/var/lib
/var/lib/libvirt
/var/lib/libvirt/images
/var/lib/libvirt/images/win10.qcow2

real	0m4.820s
user	0m4.171s
sys	0m0.643s
```
Original golang  `restic` implementation:
```
juergen@shaun:~/ghq/github.com/juergenhoetzel/pyrrhic → time restic ls latest
repository 8f93074a opened (repository version 1) successfully, password is correct
snapshot 27cb429b of [/var/lib/libvirt/images] filtered by [] at 2022-09-26 17:19:58.822747416 +0200 CEST):
/var
/var/lib
/var/lib/libvirt
/var/lib/libvirt/images
/var/lib/libvirt/images/win10.qcow2

real	0m6.954s
user	0m15.128s
sys	0m0.501s
```

**Comment by jcrist** [comment]
This fix has now been released as part of 0.9.1.
